base/src/main/java/org/bitcoinj/base/Difficulty.java

```diff
@@ -0,0 +1,190 @@
+    public static Difficulty ofCompact(long compact) {
+        long exponent = (compact >> 24) & 0xff;
+        checkArgument(exponent <= 32, () ->
+                "exponent cannot exceed 32: " + Long.toHexString(compact));
+        long mantissa = compact & 0xffffff;
+        checkArgument(mantissa <= 0x7fffff, () ->
+                "sign bit 24 cannot be set: " + Long.toHexString(compact));
+        checkArgument(mantissa >= 0x008000, () ->
+                "not optimally encoded, can shift to left: " + Long.toHexString(compact));
+        return new Difficulty(compact);
+    }
+
+    /**
+     * Construct a difficulty from an 256-bit integer value. Because the value is stored in compact form, it will
+     * likely lose precision.
+     *
+     * @param value 256-bit integer value
+     * @return constructed difficulty
+     */
+    public static Difficulty ofInteger(BigInteger value) {
+        checkArgument(value.signum() >= 0, () ->
+                "cannot be negative: " + value.toString(16));
+        checkArgument(value.compareTo(MAX_INTEGER_VALUE) <= 0, () ->
+                "too high: " + value.toString(16));
+        return new Difficulty(ByteUtils.encodeCompactBits(value));
+    }
+
+    /**
+     * Inside a block the difficulty target is represented using a compact form, sometimes called "nBits".
+     *
+     * @return difficulty target as a long in compact form
+     */
+    public long compact() {
+        return compact;
+    }
+
+    /**
+     * Returns the difficulty target as a 256 bit value that can be compared to a SHA-256 hash.
+     *
+     * @return difficulty target as 256-bit integer value
+     */
+    public BigInteger asInteger() {
+        return ByteUtils.decodeCompactBits(compact);
+    }
+
+    /**
+     * Adjust this difficulty so that actual time between blocks closely matches our target. This method doesn't take
+     * hash rate changes between intervals into account.
+     *
+     * @param actualTimespan the actual duration of the last block interval, according to the headers
+     * @param targetTimespan the duration of block intervals we're targetting at
+     * @param maxTarget      make sure adjusted difficulty doesn't get any easier than this value
+     * @return adjusted difficulty to be used for the next interval
+     */
+    public Difficulty adjust(Duration actualTimespan, Duration targetTimespan, Difficulty maxTarget) {
+        checkArgument(!actualTimespan.isNegative(), () -> "cannot be negative: " + actualTimespan);
+        checkArgument(!actualTimespan.isZero(), () -> "cannot be zero: " + actualTimespan);
+        checkArgument(!targetTimespan.isNegative(), () -> "cannot be negative: " + targetTimespan);
+        checkArgument(!targetTimespan.isZero(), () -> "cannot be zero: " + targetTimespan);
+        BigInteger adjusted = asInteger()
+                .multiply(BigInteger.valueOf(actualTimespan.getSeconds()))
+                .divide(BigInteger.valueOf(targetTimespan.getSeconds()));
```

Comment:
This might read a little easier if you create two intermediate `BigInteger` variables for "actual" and "target" and then do the divide/multiply with them. Your call.

Reply:
Done: be2256d27